src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AsyncWorkItemQueue.cs

```diff
@@ -66,13 +66,10 @@ public void RemoveCancellationSource(object key)
                         // so don't call cancel on the source only because we are done using it.
                         _cancellationMap.Remove(key);
 
-                        if (!HasAnyWork_NoLock)
-                        {
-                            Contract.Requires(_cancellationMap.Count == 0);
-
-                            // last work is done.
-                            _progressReporter.Stop();
-                        }
+                        // every works enqueued by "AddOrReplace" will be processed
+                        // at some point, and when it is processed, this method will be called to mark
+                        // work has been done.
```

Comment:
:memo: This condition is extremely difficult to verify in code.

Reply:
let me think a way to show it more explicit.

Reply:
I looked a bit not sure how to make it clear in code. so, I just added more comments listing expected order of events.